### PR TITLE
[WIP] Testable tasks

### PR DESCRIPTION
**Pull request by wardi**
Work towards factoring out testable functions from Environment methods. Some bug fixing along the way too.

### CODE REVIEW

**Comment by JackMc** [comment]
:+1: Looks great. My only worry is how we test the file-io bits.

**Comment by dborzov** [comment]
:+1: great! all these tasks don't make a lot of sense as the environment's methods by now anyhow. Not with sites as an entity now.

File: datacats/task.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    cp.add_section('passwords')
+    for n in sorted(passwords):
+        cp.set('passwords', n.lower(), passwords[n])
+
+    # Write to the sitedir so we maintain separate passwords.
+    with open(sitedir + '/passwords.ini', 'w') as config:
+        cp.write(config)
+
+
+def save(name, datadir, srcdir, ckan_version, deploy_target=None,
```
Comment (dborzov) [inline datacats/task.py]: should `save` be renamed to `save_environment` here to match `save_site` naming?

Comment (JackMc) [inline datacats/task.py]: :+1: This is a confusing name!

**Comment by wardi** [comment]
@dborzov yep, it might be better to have Environment and Site classes as separate things. This should be easier after pulling this code out.

**Comment by wardi** [comment]
:heart_eyes_cat: pylint

**Comment by JackMc** [comment]
I dunno if this is in line with your vision for this module, but perhaps creating environments/sites could be abstracted into a "task"?

The magic sequence of calls required to do so is duplicated in all three of our datacats-related projects.

**Comment by JackMc** [comment]
Alternatively this could be done by adding a "quiet" parameter to create_environment. Nvm, I think that that is definitely a better solution. :) I'll open my own PR

**Comment by wardi** [comment]
@JackMc yes, my plan is to merge this then look at changing the interface that Environment provides something actually nice to use. The only reason it's separated right now is for the progress indication, which could be handled by passing a progress_callback to a single "create my site" call.

**Comment by JackMc** [comment]
I was thinking the same :) Just had the thought while I was writing multisite and cause Dima had an issue with not calling one method and silent fails

**Comment by wardi** [comment]
Merging this now, the next tasks I'll be separating involve \larger changes to how we're running docker commands, and I don't want to hold up unit test development.

**Comment by JackMc** [comment]
Wait wait wait test on Mac it seemed broken

**Comment by JackMc** [comment]
Trying it now

**Comment by wardi** [comment]
@JackMc sorry, should have waited for your review

**Comment by JackMc** [comment]
It's ok I was caught up in multisite :stuck_out_tongue: 

Here's the traceback, looking into it now:
```
Creating environment "testenv12/primary"..........
----------------------------------------
datacats: problem was encountered:
Something that should not have happened happened when attempting to run this command:
     datacats create testenv12

It is seems to be a bug.
Please report this issue to us by creating an issue ticket at

    https://github.com/datacats/datacats/issues

so that we would be able to look into that and fix the issue.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Please quote this traceback when reporting this error:
Traceback (most recent call last):
  File "/Users/jackmccracken/sources/datacats/datacats/cli/main.py", line 92, in main
    return command_fn(opts)
  File "/Users/jackmccracken/sources/datacats/datacats/cli/create.py", line 53, in create
    datapusher=not opts['--no-datapusher'],
  File "/Users/jackmccracken/sources/datacats/datacats/cli/create.py", line 89, in create_environment
    fn()
  File "/Users/jackmccracken/sources/datacats/datacats/environment.py", line 307, in create_install_template_skin
    self.install_package_develop('ckanext-' + self.name + 'theme')
  File "/Users/jackmccracken/sources/datacats/datacats/environment.py", line 711, in install_package_develop
    stream_output=stream_output
  File "/Users/jackmccracken/sources/datacats/datacats/environment.py", line 726, in user_run_script
    stream_output=stream_output
  File "/Users/jackmccracken/sources/datacats/datacats/environment.py", line 759, in run_command
    commit=True, stream_output=stream_output)
  File "/Users/jackmccracken/sources/datacats/datacats/docker.py", line 200, in web_command
    raise WebCommandError(command, c['Id'][:12], logs)
WebCommandError:     Command: /scripts/run_as_user.sh /scripts/run.sh /project/ckanext-testenv12theme
    Docker Error Log:
    Obtaining file:///project/ckanext-testenv12theme
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/pip/commands/install.py", line 278, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/pip/req.py", line 1139, in prepare_files
    _make_build_dir(self.build_dir)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/pip/req.py", line 1527, in _make_build_dir
    os.makedirs(build_dir)
  File "/usr/lib/ckan/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/usr/lib/ckan/build'

Storing debug log for failure in /project/.pip/pip.log

----------------------------------------
```

**Comment by wardi** [comment]
@JackMc probably failing to mount the venv from its volume container

**Comment by JackMc** [comment]
Yeah, you were right. 

It was a problem where the fix permissions script wasn't hitting the virtualenv cause it wasn't mounted.

**Comment by JackMc** [comment]
Hotfix is on master, waiting to see if it passes CircleCI

**Comment by JackMc** [comment]
Answer: No, no it doesn't

**Comment by JackMc** [comment]
Working now
Penguin

**Pull request merged**